### ISSUE TITLE
[@typescript-eslint/no-inferrable-types] Misleading rule naming

### ISSUE THREAD

**Issue by higherkinded** [typescript-eslint/typescript-eslint#1673]
Not sure if it can be considered a proper issue but this rule title is misleading. TypeScript can infer pretty much everything that deals with concrete values, so the title suggests that same-signature cases are rejected, but actually they are allowed. I've checked the rule code for additional context but it turns out that all it does is just a plain `switch` that checks for primitives. Additionally, out of all structures present in the specification, it checks for... RegExp. I'm not sure why. Having stated this, I'll provide two suggestions (that may be considered as a part of some future major) which I myself find sensible:

Suggestion 1: Rename the rule to `"no-primitive-signatures"` or something similar.

Suggestion 2: Check for same-signature assignments instead of `switch`ing. That's probably going to be harder than just taking the first route but I'm willing to help if such a change makes sense for maintainers.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-inferrable-types": "on"
  }
}
```

```TS
// Anything more complex than primitive but inferrable by TypeScript
const ref: React.RefObject<HTMLElement> = useRef(null);
```

**Expected Result**
Failure

**Actual Result**
Not a word

**Comment by bradzacher** [comment]
> Additionally, out of all structures present in the specification, it checks for... RegExp. I'm not sure why.

Regex literals are encoded very clearly into the AST, so it's trivial to detect it.
Same reason that all of the primitive types are checked - the literal is encoded in the AST, so it's trivial to detect it.

---

Nitpick, but your example is code is actually wrong, as `useRef(null)` returns `React.RefObject<null>`, so the hypothetical enhancements still wouldn't error in that case for two reasons:
1) `null !== HTMLElement`, so the type annotation is required to declare the generic type parameter.
2) `useRef` actually returns `MutableRefObject`, so the type annotation is required to cast the type from the mutable type to the readonly type.

---

I think the misunderstanding here is the word `inferrable`. You're assuming it just to mean that the type is inferrable by typescript.

In this instance it's actually mainly referring to being inferrable by a human.

Some examples:

`const x: number = 1` is entirely inferrable: it's trivially inferrable by a human and also by TS.

`const ref: React.RefObject<HTMLElement> = useRef<HTMLElement>(null)` is not inferrable by either a human or TS, because of reason (2) above.

`const ref: React.MutableRefObject<HTMLElement> = useRef<HTMLElement>(null)` is trivially inferrable by TS, but I would argue that it's not trivially inferrable by a human.
This being inferrable relies upon two things:
1) the human must have prior knowledge of React's type declarations
2) the file must not declare another instance of `useRef` which returns a different type, i.e.
    ```ts
    import React from 'react';
    
    declare function useRef(arg: null): { current: HTMLElement, someOtherProp: string };
    
    function MyComponent() {
      const ref: React.RefObject<HTMLElement> = useRef(null);
    }
    ```

Both of these are not guaranteed, and thus I would say you cannot make the blanket statement that the type is inferrable, so neither can the rule.

This is why the rule behaves as it does.

---

As an aside, it can be really hard to tell when a type annotation is unnecessary outside of certain simple cases.

Simple cases include these two cases:
- checking any primitive types,
- checking type references when the types are the same (i.e it's easy to check for `const x: Foo = new Foo();`),

These are checkable, as typescript maintains a single instance of the types in the data structures, meaning you can do a simple `typeAnnotationType === valueType` check for them.
If the type uses generic type parameters, then you also need to ensure they are all equal as well, but that recurses into the same problem.

Things fall quickly into the realm of really, _really_ hard beyond those two cases, however.
```ts
type T = { prop: string };
const x: T = { prop: 'foo' };
```
Technically, this type annotation is inferrable, and unnecessary.
But even though the types are exactly equivalent, there's no simple way to compare them, beyond ofc iterating all of the properties (recursively) and ensuring that the shapes are the same.

But then this begs the question is - is it inferrable if the types are "duck type  equivalent"? Or is it only inferrable if the types are the same?
What is the definition of inferrable the rule should use?
